woocommerce/woocommerce-ios
# Remove Exclusion of Future Orders

Closes #3203.  
Fixes #2202. 
Closes #2805. 

Before | After 
--------|-------
  <img src="https://user-images.githubusercontent.com/198826/100154322-a089f300-2e62-11eb-9e92-be68d9f7368b.png" width="300">   |       <img src="https://user-images.githubusercontent.com/198826/100154095-4f79ff00-2e62-11eb-81c8-1937cda01379.png" width="300">

## Findings

We currently don't show orders that are dated in the future in the All Orders. A few months ago, we decided to show all the orders, regardless of what their `dateCreated` value is. This was done on Android (https://github.com/woocommerce/woocommerce-android/pull/2413) but I totally forgot to do the same on iOS. This fixes my mistake. 

Removing the exclusion of future orders also fixes #2202. This actually makes me wonder whether the `predicates` with the date constraint is why we're getting reports from users saying that their orders are not loading. 

https://github.com/woocommerce/woocommerce-ios/blob/1c94af3a701bd16bcbadd142f08acd50cd421666/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L140-L145

We never update the date argument in this constraint so if the app is never terminated, it's possible that the constraint will stay on the same date. For example:

Scenario | Date | Predicate | Result
--------|-------|---|---
User installs and opens the app.        |    2020-11-20  | `dateCreated < 2020-11-21` | All good and all orders (except future orders) should be loaded
User comes back to the app 2 days later. The app was never terminated by iOS. | 2020-11-22 | `dateCreated < 2020-11-21` | Orders submitted on Nov 21 and 22nd are not shown. 

🤦 

## Changes

This removes the exclusion of future orders, which also fixes the bug I mentioned above. 

In addition to that, I copied the [`OrdersViewModelTests`](https://github.com/woocommerce/woocommerce-ios/blob/issue/2202-fix-orders-predicate/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift) to its newer iOS-13 only version, [`OrderListViewModelTests`](https://github.com/woocommerce/woocommerce-ios/blob/issue/2202-fix-orders-predicate/WooCommerce/WooCommerceTests/ViewRelated/Orders/OrderListViewModelTests.swift). This should allow us to safely remove the deprecated `OrdersViewModel` when we drop iOS 12.0 support. 

## Testing

1. Submit a new order on the web.
2. Set the device's date back one month.
3. Run the app on the device.
4. Navigate to All Orders. 
5. Confirm that you still see the order you just submitted. It should be under the “Upcoming” section.
6. Set the device date back to today.
7. Pull-to-refresh to All Orders. 
8. Confirm that all the orders are still shown. But the order you submitted should be in the “Today” section. 
9. Navigate away from the All Orders tab. Maybe go to Processing.
10. Submit another order on the web. 
11. Navigate back to the All Orders tab. 
12. Don't do a pull-to-refresh. Wait a while and then confirm that the last order you submitted is shown. 

## Submitter Checklist
